indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java

```diff
@@ -184,275 +136,393 @@ public String getType()
+    } else {
+      firehoseFactory = new ReplayableFirehoseFactory(
+          delegateFirehoseFactory,
+          ingestionSchema.getTuningConfig().isReportParseExceptions(),
+          null,
+          null
+      );
+    }
 
-    final TaskLock myLock = Iterables.getOnlyElement(getTaskLocks(toolbox));
-    final Set<DataSegment> segments = Sets.newHashSet();
+    try {
+      final Map<Interval, List<ShardSpec>> shardSpecs = determineShardSpecs(toolbox, firehoseFactory);
 
-    final Set<Interval> validIntervals = Sets.intersection(granularitySpec.bucketIntervals().get(), getDataIntervals());
-    if (validIntervals.isEmpty()) {
-      throw new ISE("No valid data intervals found. Check your configs!");
-    }
+      final String version;
+      final DataSchema dataSchema;
+      if (!determineIntervals) {
+        version = Iterables.getOnlyElement(getTaskLocks(toolbox)).getVersion();
+        dataSchema = ingestionSchema.getDataSchema();
+      } else {
+        Interval interval = JodaUtils.umbrellaInterval(shardSpecs.keySet());
+        TaskLock lock = toolbox.getTaskActionClient().submit(new LockAcquireAction(interval));
+        version = lock.getVersion();
+        dataSchema = ingestionSchema.getDataSchema().withGranularitySpec(
+            ingestionSchema.getDataSchema()
+                           .getGranularitySpec()
+                           .withIntervals(
+                               JodaUtils.condenseIntervals(
+                                   shardSpecs.keySet()
+                               )
+                           )
+        );
+      }
 
-    for (final Interval bucket : validIntervals) {
-      final List<ShardSpec> shardSpecs;
-      if (targetPartitionSize > 0) {
-        shardSpecs = determinePartitions(bucket, targetPartitionSize, granularitySpec.getQueryGranularity());
+      if (generateAndPublishSegments(toolbox, dataSchema, shardSpecs, version, firehoseFactory)) {
+        return TaskStatus.success(getId());
       } else {
-        int numShards = ingestionSchema.getTuningConfig().getNumShards();
-        if (numShards > 0) {
-          shardSpecs = Lists.newArrayList();
-          for (int i = 0; i < numShards; i++) {
-            shardSpecs.add(new HashBasedNumberedShardSpec(i, numShards, null, jsonMapper));
-          }
-        } else {
-          shardSpecs = ImmutableList.<ShardSpec>of(NoneShardSpec.instance());
-        }
+        return TaskStatus.failure(getId());
       }
-      for (final ShardSpec shardSpec : shardSpecs) {
-        // ShardSpec to be published.
-        final ShardSpec shardSpecForPublishing;
-        if (ingestionSchema.getTuningConfig().isForceExtendableShardSpecs()) {
-          shardSpecForPublishing = new NumberedShardSpec(
-              shardSpec.getPartitionNum(),
-              shardSpecs.size()
-          );
-        } else {
-          shardSpecForPublishing = shardSpec;
+    }
+    finally {
+      if (firehoseFactory instanceof Closeable) {
```

Comment:
Hm, the purpose of closing the factory is to delete the temporary cache files. Technically, this doesn't really need to be done, since we would expect the `deleteOnExit()` mechanism to handle cleanup when the process terminates. I added it since I thought it would be nice to clean up as soon as possible, but having a Closeable FirehoseFactory (understandably) seems to be adding a lot of uncertainty. I'll rework it to just clean up the temp files when the process terminates using `deleteOnExit()`, which should happen at pretty much the same time in indexing tasks.